operator/controllers/common/providers/dynatrace/client/client.go

```diff
@@ -0,0 +1,242 @@
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"k8s.io/klog/v2"
+)
+
+const (
+	defaultAuthURL                  = "https://sso-dev.dynatracelabs.com/sso/oauth2/token"
+	oAuthGrantType                  = "grant_type"
+	oAuthGrantTypeClientCredentials = "client_credentials"
+	oAuthScope                      = "scope"
+	oAuthClientID                   = "client_id"
+	oAuthClientSecret               = "client_secret"
+)
+
+// OAuthScope represents a scope provided for the registered OAuth client interacting with the DT API
+type OAuthScope string
+
+// These constants define the scopes that we currently need for the DQL metric functionality. This list might extend as new features will be added.
+// For now, we keep this at the minimum set of scopes required, as these are currently likely to change
+const (
+	OAuthScopeStorageMetricsRead    = "storage:metrics:read"
+	OAuthScopeEnvironmentRoleViewer = "environment:roles:viewer"
+)
+
+type OAuthResponse struct {
+	AccessToken string `json:"access_token"`
+}
+
+type apiConfig struct {
+	serverURL        string
+	authURL          string
+	oAuthCredentials oAuthCredentials
+}
+
+type APIConfigOption func(config *apiConfig)
+
+func WithAuthURL(authURL string) APIConfigOption {
+	return func(config *apiConfig) {
+		config.authURL = authURL
+	}
+}
+
+// WithScopes passes the given scopes to the client config
+func WithScopes(scopes []OAuthScope) APIConfigOption {
+	return func(config *apiConfig) {
+		config.oAuthCredentials.scopes = scopes
+	}
+}
+
+// NewAPIConfig returns a new apiConfig that can be used for initializing a DTAPIClient with the NewAPIClient function
+func NewAPIConfig(serverURL string, secret string, opts ...APIConfigOption) (*apiConfig, error) {
```

Comment:
would it make sense to move the apiConfig and other stuff to a separate file? For example we will have:

apiClient.go
apiConfig.go
oauthCredentials.go

all part of the same package, but in separate files so it's easier to read and orientate through the code

commonly used methods/constants can be moved to common.go